backend/src/main/java/fk/prof/backend/model/assignment/impl/ProcessGroupAssociationStoreImpl.java

```diff
@@ -0,0 +1,43 @@
+package fk.prof.backend.model.assignment.impl;
+
+import fk.prof.backend.model.assignment.*;
+import recording.Recorder;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+
+public class ProcessGroupAssociationStoreImpl implements ProcessGroupAssociationStore, ProcessGroupDiscoveryContext {
+  private final Map<Recorder.ProcessGroup, ProcessGroupDetail> processGroupLookup = new ConcurrentHashMap<>();
+  private final int thresholdForDefunctRecorderInSecs;
+
+  public ProcessGroupAssociationStoreImpl(int thresholdForDefunctRecorderInSecs) {
+    this.thresholdForDefunctRecorderInSecs = thresholdForDefunctRecorderInSecs;
+  }
+
+
+  //Called by backend daemon thread when load is reported to leader and leader responds with associations
+  @Override
+  public void updateProcessGroupAssociations(Recorder.ProcessGroups processGroups, BiConsumer<ProcessGroupContextForScheduling, ProcessGroupAssociationResult> postUpdateAction) {
+    //Remove process group associations which are not returned by leader
+    for(Recorder.ProcessGroup processGroup: this.processGroupLookup.keySet()) {
+      if(!processGroups.getProcessGroupList().contains(processGroup)) {
+        ProcessGroupDetail processGroupDetail = this.processGroupLookup.remove(processGroup);
+        postUpdateAction.accept(processGroupDetail, ProcessGroupAssociationResult.REMOVED);
+      }
+    }
+
+    //Add process group associations which are returned by leader
+    for(Recorder.ProcessGroup processGroup: processGroups.getProcessGroupList()) {
+      ProcessGroupDetail existingValue = this.processGroupLookup.putIfAbsent(processGroup, new ProcessGroupDetail(processGroup, thresholdForDefunctRecorderInSecs));
```

Comment:
put if absent is required here. Leader always returns all the process groups associated with a backend. Process groups which are already present in the lookup should not be associated with a new processgroupdetail instance. Only the ones which are new. Hence the putIfAbsent semantic